### PR TITLE
Add gh actions for building and testing the repo

### PR DESCRIPTION
**Pull request by vzaidman**
Introduced config that triggers GH actions that build, lint, and test the repo.

These are triggered when a pull request is created or it's latest commit is updated.

If a new job is triggered, the previous is abandoned.

* Please notice that tests are currently disabled for windows because they fail. This will be fixed in a subsequent diff.

Test:
* See the Github actions on this PR: 
Screenshot 2024-08-02 at 17 14 07

* Example of a run cancelled by pushing a new commit before the current is processed by Github actions: https://github.com/facebook/metro/actions/runs/10216106999

### CODE REVIEW

File: .github/workflows/build-and-test.yml [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+    - uses: actions/checkout@v4
+    - uses: ./.github/actions/yarn-install
+    - run: yarn test-coverage
+    - run: "./.github/scripts/install_codecov.sh"
+    - run: "./codecov -t ${{ secrets.CODECOV_TOKEN }} -f ./coverage/coverage-final.json"
+
+  test:
+    strategy:
+      matrix:
+        runs-on: ['ubuntu-latest'] # 'windows-latest'
```
Comment (robhogan) [inline .github/workflows/build-and-test.yml:42]: IIRC we have a bunch of tests currently not passing on Windows unfortunately and this probably can't be enabled on `main` in the short term, but in CircleCI we have the convention of [`windows` branches to trigger CI on Windows](https://github.com/facebook/metro/blob/e3aaa0bac2d91598646341dfd5621c28cdf7370f/.circleci/config.yml#L133), which is useful if we're looking at a specific test suite.

Alternatively... if we could always run on Windows but have Jest automatically skip known-broken Windows tests *only* on Windows - maybe with a [`--filter`](https://jestjs.io/docs/cli#--filterfile) file, that'd be a much better setup.

File: package.json [hunk 2]
```diff
@@ -94,5 +94,8 @@
   "dependencies": {},
   "engines": {
     "node": ">=18"
+  },
+  "volta": {
+    "node": "18.0.0"
```
Comment (robhogan) [inline package.json:99]: Is this related to the PR somehow?

File: .github/workflows/build-and-test.yml [hunk 3]
```diff
@@ -0,0 +1,52 @@
+name: facebook/metro/build-and-test
+on:
+  workflow_call:
+  pull_request:
+    types: [opened, synchronize]
+
+# head_ref is per PR, so this ensures that updaing the latest PR commit
```
Comment (robhogan) [inline .github/workflows/build-and-test.yml:7]:
```suggestion
# head_ref is per PR, so this ensures that updating the latest PR commit
```

File: .github/workflows/build-and-test.yml [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+  cancel-in-progress: true
+
+defaults:
+  run:
+    shell: bash
+
+jobs:
+  run-js-checks:
+    runs-on: ubuntu-latest
+    name: "Typescript, Link, Smoke Test"
```
Comment (robhogan) [inline .github/workflows/build-and-test.yml:20]:
```suggestion
    name: "TypeScript, Lint, Smoke Test"
```

File: .github/actions/yarn-install/action.yml [hunk 5]
```diff
@@ -0,0 +1,23 @@
...
+      - 'min-supported'
+
+runs:
+  using: "composite"
+  steps:
+  - uses: actions/setup-node@v4
+    with:
+      # if both node-version-file and node-version are provided, node-version is used
+      node-version-file: 'package.json'
+      node-version: ${{ inputs.node-version-type == 'lts' && 'lts/*' || ''  }}
```
Comment (robhogan) [inline .github/actions/yarn-install/action.yml:18]: I don't fully understand the use of `node-version-file`/`node-version` here - can we not use a matrix like the example at https://docs.github.com/en/actions/writing-workflows/choosing-what-your-workflow-does/using-a-matrix-for-your-jobs#example-using-a-single-dimension-matrix ?

**Review by robhogan — COMMENTED** [review]
Thanks for taking this on! Just a couple of nits/questions

File: .github/workflows/build-and-test.yml (hunk 1 above)
Comment (vzaidman) [reply .github/workflows/build-and-test.yml:42]: I was planning to fix the windows tests on a next commit and will enable it. WDYT?

File: package.json (hunk 2 above)
Comment (vzaidman) [reply package.json]: Yes, this tells on which version of node tests will run.
Without it, setting `node-version-file: 'package.json'` will take version `20.2` based on the engine set to `"node": ">=18"`.
But I'd like to test the minimum node version we support and LTS instead.
See the improved comment on `.github/workflows/build-and-test.yml` below.

File: .github/actions/yarn-install/action.yml (hunk 5 above)
Comment (vzaidman) [reply .github/actions/yarn-install/action.yml:25]: I've updated the comment. please see if you agree with this approach. I can certainly hardcode the versions, but then the minimum supported version on `package.json` will be maintained separately from that hardcoded version in the pipelines, which is not the end of the world, but with no single point of truth they might diverge.

Comment (robhogan) [reply .github/actions/yarn-install/action.yml:25]: My main concern here is the CI configuration effectively leaking into the repo (the `volta` field) in a way that's not obvious if you're looking at it from the `package.json` side. And, if you're a contributor actually using Volta, this is effectively breaking by forcing you to develop using 18.0.0, right?

> with no single point of truth they might diverge.

This is a fair point but I don't see it as a disaster:

 - Bumping the CI minimum without bumping Metro's minimum isn't a concern, I can't see a reason we'd ever do that voluntarily.
 - If we bumped Metro's minimum but forgot the CI minimum, the worst that happens is CI breaks some time later when we actually break compatibility, so we notice and update it. 
 - We probably *should* have an E2E test that builds and publishes, then installs Metro using something like Verdaccio, like RN does - then we'd fail fast because the `engines` requirement would cause installation to fail.
 
There are upsides to hardcoding the matrix - it's cleaner and more GHA-native (automatically reflected in test names, etc). I think that'd be my preference tbh. `['current', 'lts/*', 'lts/-1', '18.0']` would give us good coverage.

Comment (vzaidman) [reply .github/actions/yarn-install/action.yml]: gotcha. I also spoke with @huntie about that.
could you please take a look at the updated code?

File: package.json (hunk 2 above)
Comment (vzaidman) [reply package.json]: removed this change.

File: .github/actions/yarn-install/action.yml [hunk 6]
```diff
@@ -0,0 +1,27 @@
...
+  current-node-version:
+    required: false
+    default: '20.2'
+
+runs:
+  using: "composite"
+  steps:
+  - uses: actions/setup-node@v4
+    with:
+      node-version: ${{ inputs.node-version-type == 'current' && inputs.current-node-version || inputs.min-supported-node-version }}
```
Comment (robhogan) [inline .github/actions/yarn-install/action.yml:22]: I'm not sure what value this `node-version-type` indirection gives us - why not pass `node-version` through directly from the matrix, and then we can set it once at the top level?

File: .github/workflows/build-and-test.yml (hunk 1 above)
Comment (robhogan) [reply .github/workflows/build-and-test.yml:41]: Afaik fixing tests on Windows isn't trivial but yeah we can follow up. I'd probably remove the commented-out `'windows-latest'` for now though.

File: .github/workflows/build-and-test.yml [hunk 7]
```diff
@@ -0,0 +1,51 @@
...
+    - run: yarn test-coverage
+    - run: "./.github/scripts/install_codecov.sh"
+    - run: "./codecov -t ${{ secrets.CODECOV_TOKEN }} -f ./coverage/coverage-final.json"
+
+  test:
+    strategy:
+      matrix:
+        runs-on: ['ubuntu-latest'] # 'windows-latest'
+        node-version-type: ['current', 'min-supported']
+    name: "Tests [${{ matrix.runs-on }} machine, ${{ matrix.node-version-type }} Node.js]"
```
Comment (huntie) [inline .github/workflows/build-and-test.yml:43]: nit, also ideally let's print the numeric Node version, e.g. `Tests [Node.js 20, ubuntu-latest]`

```suggestion
    name: "Tests [Node.js ${{ matrix.node-version-type }}, ${{ matrix.runs-on }}]"
```

File: .github/workflows/build-and-test.yml [hunk 8]
```diff
@@ -0,0 +1,51 @@
...
+
+# head_ref is per PR, so this ensures that updating the latest PR commit
+# will cancel the previous run of the workflow and trigger a new one
+concurrency:
+  group: "build-and-test-${{ github.head_ref }}"
+  cancel-in-progress: true
+
+defaults:
+  run:
+    shell: bash
```
Comment (huntie) [inline .github/workflows/build-and-test.yml:14]: Q: Do we have any compatibility issues with zsh (assuming as default)? I'd imagine everything we'd need would work, since zsh is a superset of Bash.

File: .github/workflows/build-and-test.yml [hunk 9]
```diff
@@ -0,0 +1,51 @@
...
+
+jobs:
+  run-js-checks:
+    runs-on: ubuntu-latest
+    name: "Typescript, Lint, Smoke Test"
```
Comment (huntie) [inline .github/workflows/build-and-test.yml:19]: nit: This is _Flow_, TypeScript, Lint ... — can we genericise?
```suggestion
    name: "Static analysis"
```

or "Lint", "Run JS checks"

File: .github/workflows/build-and-test.yml [hunk 10]
```diff
@@ -0,0 +1,51 @@
...
+      - uses: actions/checkout@v4
+      - uses: ./.github/actions/yarn-install
+      - run: yarn typecheck
+      - run: yarn typecheck-ts
+      - run: yarn lint
+      - run: yarn test-smoke
+
+  test-with-coverage:
+    runs-on: ubuntu-latest
+    name: "Test Coverage Calculation"
```
Comment (huntie) [inline .github/workflows/build-and-test.yml:30]: nit (all tests need to run to calculate coverage)
```suggestion
    name: "Tests with coverage"
```

File: .github/actions/yarn-install/action.yml [hunk 11]
```diff
@@ -0,0 +1,27 @@
...
+    type: choice
+    options:
+      - 'current'
+      - 'min-supported'
+  min-supported-node-version:
+    required: false
+    default: '18.12'
+  current-node-version:
+    required: false
+    default: '20.2'
```
Comment (huntie) [inline .github/actions/yarn-install/action.yml:15]: I reckon we can reduce to `18.x`, `20.x` as specifiers at the GitHub level so we can simplify this. Then we could receive one `node-version` arg as a string, and more easily test 3 versions in future if we need to.

File: .github/workflows/build-and-test.yml (hunk 9 above)
Comment (robhogan) [reply .github/workflows/build-and-test.yml:19]: This used to be `Type check, lint, smoke test`, which is probably more accurate tbh. But yeah, it might be cleaner if we separated the smoke test from static analysis.

File: .github/workflows/build-and-test.yml (hunk 7 above)
Comment (robhogan) [reply .github/workflows/build-and-test.yml:43]: I'm in favour of eliminating `node-version-type` and just using a `node-version` matrix, which still supports aliases like `lts/*`, `lts/-n`, `current`.

File: .github/actions/yarn-install/action.yml (hunk 11 above)
Comment (robhogan) [reply .github/actions/yarn-install/action.yml:15]: I would like to have a test on `18.0` though since we claim to support it, and there are some differences vs later minors. `['current', 'lts/*', 'lts/-1', '18.0']` (which resolves to 21.x, 20.x, 18.x, 18.0), IMO.

File: .github/actions/yarn-install/action.yml (hunk 6 above)
Comment (vzaidman) [reply .github/actions/yarn-install/action.yml:22]: It certainly can be done but I wanted to avoid the different workflows using the action (true there's only one at this point) from passing their own `node-version` to avoid setting it separately in different files.

File: .github/workflows/build-and-test.yml (hunk 8 above)
Comment (vzaidman) [reply .github/workflows/build-and-test.yml]: Unfortunately zsh is not installed on ubuntu images (and installing it is +20sec to each run)
We can use macOS machines instead, but even there zsh is not "officially supported" but only supported as a "custom shell" running a template: https://github.com/actions/runner-images/issues/264 
I still tested it to ensure it doesn't work out of the box, and it indeed doesn't:
https://github.com/facebook/metro/actions/runs/10286164783

File: .github/actions/yarn-install/action.yml (hunk 11 above)
Comment (vzaidman) [reply .github/actions/yarn-install/action.yml]: Sounds good to me, however @huntie disagreed that it makes sense to test the minimum and lts versions before.
WDYT @huntie?

Comment (vzaidman) [reply .github/actions/yarn-install/action.yml]: BTW @robhogan, `current` resolves to `v22.6.0`.
https://github.com/facebook/metro/actions/runs/10286511772/job/28467353526

Comment (robhogan) [reply .github/actions/yarn-install/action.yml]: Yeah that's fair - essentially it's unstable (though fwiw these CI signals aren't land blocking yet). I'd like to test 18.0 (which should be stable), I'd be OK to use `22.x`, `20.x` instead of the aliases.

It *would* still be good to know asap if we're incompatible with a new Node major, because we don't put a "maximum" constraint on `engines` so Node releases can break users (it's [happened before](https://github.com/facebook/metro/pull/752)), but ideally we'd run that as a separate scheduled (eg nightly) run rather than anything that potentially signals a diff-time failure unrelated to the diff. We can follow up with that.

Comment (vzaidman) [reply .github/actions/yarn-install/action.yml]: OK. switched to `['18.0', '18.x', '20.x']` and added to my TODO list to introduce a nightly job that tests `lts/*` and `current`.

**Comment by robhogan** [comment]
@vzaidman LGTM, thanks for iterating - if you want to import I *think* the diff should now be auto-stamped because I've approved it here (never actually tried that tbh!).

File: .github/actions/yarn-install/action.yml (hunk 11 above)
Comment (robhogan) [reply .github/actions/yarn-install/action.yml]: One more for the nightly TODO list then - we should have a test that doesn't use (maybe deletes) `yarn.lock` and pulls the latest versions of all our dependencies, so we can see if a dependency update breaks us somehow.

Comment (vzaidman) [reply .github/actions/yarn-install/action.yml]: np

Comment (vzaidman) [reply .github/actions/yarn-install/action.yml]: @huntie please approve as well

File: .github/workflows/build-and-test.yml [hunk 12]
```diff
@@ -0,0 +1,51 @@
...
+      - uses: ./.github/actions/yarn-install
+      - run: yarn typecheck
+      - run: yarn typecheck-ts
+      - run: yarn lint
+      - run: yarn test-smoke
+
+  test-with-coverage:
+    runs-on: ubuntu-latest
+    name: "Tests with coverage"
+    steps:
```
Comment (huntie) [inline .github/workflows/build-and-test.yml:31]: Wait, this indentation looks wrong. Can we triple-check this, and enable Prettier on YAML files in our repo?

Comment (vzaidman) [reply .github/workflows/build-and-test.yml:31]: We already have validation on yml.

Both idents are valid in yml apparently: https://stackoverflow.com/questions/17014460/yaml-indentation-for-array-in-hash
Also, evidently it works.

However, I'll make sure they are formatted the same across the file regardless. Thx for noticing.

Comment (huntie) [reply .github/workflows/build-and-test.yml:31]: Yes, but can we please enable auto-formatting? 🥹 Good to go 👍🏻
